Refactor the command abstraction

nrc opened the pull request:
This is a major refactoring to the client's internals. The main point of the refactoring is to replace the old KvRequest trait and associated code with an a cleaner abstraction. The new abstraction is a plan, which is a strategy for executing a request. Plans are composable and more self-contained. They are also more flexible and in some cases more efficient than the old system.

As well as the major refactoring (in the last commit) there are a bunch of smaller, tidying-up refactorings in earlier commits.

Closes #205, #206. It is a prerequisite for #111, #208, and #209.

PTAL @andylokandy @ekexium

andylokandy commented on src/request/plan_builder.rs:
```diff
@@ -0,0 +1,178 @@
+    /// Apply a processing step to a response (usually only needed if the request is sent to a
+    /// single region because post-porcessing can be incorporated in the merge step for multi-region
+    /// requests).
+    pub fn post_process(self) -> PlanBuilder<PdC, ProcessResponse<P, P::Result>, Ph>
```
What does this mean?

nrc replied:
Which part needs clarification? The use case is something like you pass a `RawGetResponse` in and get a `Value` out, i.e., it is discarding some unneeded information and transforming types to something a little bit higher level.

andylokandy replied:
I mean, we can do this by default, as there will be no need to call this explicitly,

nrc replied:
I'm not sure what you mean, we do this by default, but for users who want to go a non-default path, they will need to call it.

andylokandy commented on src/request/plan_builder.rs:
```diff
@@ -0,0 +1,178 @@
+}
+
+/// Indicates that a request operates on a single key.
+pub trait SingleKey {
```
Who impls this trait?

nrc replied:
It is impl'ed for requests, e.g., `RawGetRequest: SingleKey`, it probably never needs to be impl'ed by users of the client, only implementors (i.e., we would need to add it for newly supported requests), but users of the crate might need to use it as a bound if they are doing something fancy.

andylokandy replied:
I still can't find any impl for requests by text-search. What did I miss?

ekexium replied:
@andylokandy The impls are in raw/requests.rs and transaction/requests.rs. GitHub hides them 🤦🏻‍♂️

andylokandy commented on src/store.rs:
```diff
@@ -77,36 +49,61 @@ where
 
 /// The range used for request should be the intersection of `region_range` and `range`.
-fn bound_range(region_range: (Key, Key), range: BoundRange) -> (Key, Key) {
+fn bound_range(region_range: (Key, Key), range: (Key, Key)) -> (Key, Key) {
```
Rename to `range_intersect`?

nrc commented on src/store.rs:
```diff
@@ -4,69 +4,41 @@ use crate::{pd::PdClient, BoundRange, Key, Region, Result};
 /// Maps keys to a stream of stores. `key_data` must be sorted in increasing order
-pub fn store_stream_for_keys<KeyData, IntoKey, I, PdC>(
-    key_data: I,
+pub fn store_stream_for_keys<K, K2, PdC>(
```
So that you can, e.g., pass in a KvPair and get out just the key.

andylokandy replied:
I see, so could we rename it to `Ret` or `Res`?

ekexium reviewed (commented):
I haven't comprehensively read all the changes. Mostly LGTM. Some of my first thoughts are:

In the new pattern, I think it's easy to miss some steps or put the steps in an incorrect order. Are there any methods that help avoid this?

I tried running integration tests. It failed with `thread 'crud' panicked at '`async fn` resumed after completion'` and other kinds of errors.

nrc commented:
> In the new pattern, I think it's easy to miss some steps or put the steps in an incorrect order. Are there any methods that help avoid this?

I use phantom types to ensure that a request is always targetted and that is the only step which if missed will cause a programming error. Other missed/out of order steps might cause logic errors, but it is also possible that the user wants those errors (e.g., if writing a tool to chaos test TiKV). I also tried a few ways to ensure the correct order, but they were pretty ugly and it was difficult to make them optional. I think that it is hard to make an error if you use the transaction API, and if you use the Plan API, then you get more flexibility but also more scope for errors.

TszKitLo40 commented:
Can you please resolve the conflict?
And it seems that there is also some mismatched types
![Uploading image.png…]()

nrc commented:
I think this should pass integration tests, however, it should not land until we've resolved #235.

PTAL @andylokandy @ekexium

ekexium commented on src/lib.rs:
```diff
@@ -6,6 +6,8 @@
 #![allow(clippy::type_complexity)]
 #![allow(incomplete_features)]
 #![cfg_attr(test, feature(specialization))]
```
So we won't support build on the stable channel?

ekexium commented on src/request/plan.rs:
```diff
@@ -0,0 +1,236 @@
+}
+
+#[derive(Clone)]
+pub struct MergeResponse<P: Plan, In, M: Merge<In>> {
```
Some plans are named after "xxxPlan", some are "xxxResponse". Could we use a similar pattern for the names?

ekexium reviewed (commented):
I tried 
1. `PD_ADDRS="127.0.0.1:2379" cargo test --package tikv-client --test integration_tests --features integration-tests -- --nocapture --ignored`, which writes a lot of data and forces region splitting.
2. and then `PD_ADDRS="127.0.0.1:2379" cargo test --package tikv-client --test integration_tests --features integration-tests -- --nocapture`

Then `txn_bank_transfer` failed. **I'm not sure** if it's a bug, because the ignored test is not supposed to run with transactional tests (may cause invalid region boundaries for transactional API). But in `master` it passed, so there might be some problem. 

Anyway I think we need to add some multi-region tests in CI.

The `merge` is composable while `post_process` is bound to the response type. Is it possible that we need different post_process for the same type of request in the future?

ekexium reviewed (commented):
I still get ```panicked at '`async fn`` resumed after completion', /home/eke/code/client-rust/src/pd/client.rs:104:24``` errors.
My steps to reproduce:
1. `tiup playground nightly --db 0 --tiflash 0 --monitor false`
2. The first few runs often failed with such errors.

nrc commented on src/transaction/requests.rs:
```diff
@@ -586,46 +314,73 @@ pub fn new_pessimistic_lock_request(
-        result.take_locks()
+    fn apply_shard(&mut self, shard: Self::Shard, store: &Store) -> Result<()> {
+        self.set_context(store.region.context()?);
+        self.set_mutations(shard.into_iter().map(Into::into).collect());
```
good catch!

nrc commented on src/transaction/requests.rs:
```diff
@@ -180,197 +130,67 @@ impl HasLocks for kvrpcpb::ScanResponse {
     commit_version: u64,
 ) -> kvrpcpb::ResolveLockRequest {
     let mut req = kvrpcpb::ResolveLockRequest::default();
-    req.set_context(context);
```
We only set the context as a way to pass in which store to send to, in the new system that is done explicitly, so the context hack is no longer needed

ekexium replied:
I get it

nrc commented on src/store.rs:
```diff
@@ -77,36 +49,61 @@ where
+        .boxed()
+}
+
+pub fn store_stream_for_range_by_start_key<PdC: PdClient>(
```
it almost can, but making the return type correct is tricky. I'm pretty sure it can be done with some fancy generics and a couple of closures, but life is too short

nrc commented:
One issue with the integration tests is that you must leave the servers some time to start up. If you immediately start the servers, then run integration tests, there are mysterious errors because the cluster does not pass validation. However, if you wait a while (about 15s seems to be enough on my machine), the tests complete without errors.

nrc commented:
> Anyway I think we need to add some multi-region tests in CI.

Agree.

> I noticed the merge is composable while post_process is bound to the response type. Is it possible that we may need different post_process for the same type of request in the future?

Still pondering this, otherwise, I've addressed all the feedback and tests are passing.

ekexium commented on src/pd/client.rs:
```diff
@@ -103,7 +103,9 @@ pub trait PdClient: Send + Sync + 'static {
             async move {
                 let start_key = match start_key {
-                    None => return Ok(None),
+                    None => {
```
Can remove the curly brackets

ekexium reviewed (approved):
LGTM

ekexium commented on src/request/plan.rs:
```diff
@@ -0,0 +1,303 @@
+                result = self.inner.execute().await?;
+            } else {
+                let mut clone = self.clone();
+                match clone.backoff.next_delay_duration() {
```
The `clone` is always cloned from `self`, thus the status of `backoff` is always at its initial state.
